docs: add design doc for settings extensions (#2834)
bottlerocket-os/bottlerocket: docs/design/settings/extensions/modular-settings-extensions.md

### Extension Binaries
All settings in a Bottlerocket host will be described by their own settings extension, which are defined by a binary packaged onto the system and symlinked into a common directory, as well as a configuration file placed in a corresponding `settings.d`-style directory.
Extension binaries must respond to command line arguments following a specified protocol spoken by the Bottlerocket `apiserver`.
The protocol is versioned, so all commands to an extension binary begin with the protocol version (in this document, all commands fall under the initial `proto1` protocol version -- so-named to avoid conflation with settings versions, which will be explained later).
Command line arguments are used to signify the "request" being made of the extension.
Exit codes will be used to signal the status of the request, with formatted output being delivered on `stdout`, and logs delivered on `stderr`.

As an example, suppose a user wants to set the "message of the day" for their instance:
```bash
$ apiclient set motd="Hello, Bottlerocket!"

> ```

Reviewer: How will the type inference work? Same sort of approach as [massage_set_input](https://github.com/bottlerocket-os/bottlerocket/blob/d5acb9e7b33121f0bab1f3102b2b8c295d750ed0/sources/api/apiclient/src/main.rs#L588) today?

Author: Hmm, I hadn't considered this. I think we could resolve this via the same algorithm with the same caveats. It's worth considering if we can come up with a cleaner solution here, but I don't think failing to come up with one introduces a change in behavior.

> The apiserver would load the current value of `motd` and then create a new object based on the inputted string.

Reviewer 2: I may have already asked this, but why does it need to load the current value before proceeding with the rest of this?

Author: You may be remembering an earlier form of this design where `get` requests would be routed to the extension to respond to. The idea here is that the API server can handle any arbitrary `get` request by referring to the stored JSON document.

Since the API server is also completely capable of handling arbitrary patches to JSON objects, it performs sets by changing the values requested by the user, and then just asking the settings extensions to validate them for correctness.

It would then find the extension responsible for `motd` and make a request to determine whether or not the change should proceed:

```bash
$ motd proto1 set --setting-version v1 --value '"Hello, Bottlerocket!"'  # Values are passed as JSON, hence the quoting
```

### Extension Naming & Setting Ownership

> Every top-level setting in Bottlerocket's API must be owned by a singular extension, and that extension will share a name with the setting.

Reviewer 2: I think I might want to write a single extension that handles multiple settings. Will that be possible?

Author: Can you given an example of what you'd want here? It's not possible with this design, though you could probably get pretty close to what you want by sharing code via shared rust crates and taking advantage of cross-validation.

This name will also be shared by the extension's configuration file and binary components.
This means that whenever the system must interact with a setting (be it via user action on the API, or rendering templates to be used as configuration files), Bottlerocket can identify the setting extension to invoke in order to satisfy that interaction.

As an example, the following suggests the filesystem layout for a settings extension for the `settings.network` settings:

```
usr
└── lib
    └── bottlerocket
        └── settings
            ├── config.d
            │   └── network.toml
            └── extensions.d

>                 └── network -> /usr/bin/network-extension

Reviewer: Is the 1:1 match between config file and symlink name enforced, or just a convention?

Author: It's enforced: This is how the system locates the configuration file for the given extension.

```

#### Transactional Writes and Cross-Validation
Bottlerocket today is missing a mechanism for cross-validation of settings.

> As an example, suppose we want to represent a range of integers in settings by allowing a `setting.min-value` and `setting.max-value` to be set -- it should be possible for settings extensions to set both simultaneously, while also verifying that the state from setting both is valid (`setting.min_value <= setting.max_value`).

Reviewer: Does the switch from `min-value` to `min_value` signify anything?

Author: Whoops! That's a mistake.

Let us cast this requirement in a different light: imagine an OS administrator wishes to restrict what values can be written to an existing setting -- for example, they may wish to only allow settings values to `network.hosts` so long as they include as a subset the entries mandated by their organization.
To allow this, settings extensions will be capable of registering themselves as a *validator* for any other setting on the host, including settings that they do not own.
When handling write requests for multiple settings in a single transaction, the `apiserver` will first gather the resulting writes that would occur under each settings extension involved in the writes.
Then the `apiserver` will provide the provisional settings state to each settings extension which has been registered as a validator of any of the target settings.
Any one settings extension has the power to halt the transaction by returning a non-zero exit-code on validation.

> Validations must specify the setting which triggers the validation in their configuration file:

Reviewer: nit: is this reading correct?
```suggestion
Extensions must specify the setting which triggers the validation in their configuration file:
```

```toml
# auditor.toml -- setting extension config file for `settings.auditor`
[extension.validates]

> network = "v1"

Reviewer: Are extensions allowed to validate top-level settings that don't exist? E.g.,
```
[extension.validates]
not-found = "v1"
```

What about versions that the extension doesn't support?

Author: Good questions. I'll update the doc, but the answer to both is no.

At the spec/packaging level, you'd need to specify a dependency on settings that are required. We can add a build-time check to ensure that validated versions are supported.

At runtime either of these will result in an error occurring and the setting failing to be set.

Author: I've changed my mind about this. I think the first case is still valid. If you attempt to validate a setting which isn't on the system, then that will silently succeed.

This could happen in cases where a vendor creates a settings extension that you are interested in, and it takes on many responsibilities but you don't particularly care for the validation functionality.

A version not being present will cause validation to fail. In this case, the author clearly wanted a validation to occur against settings which are present on the system but it cannot succeed.

Settings should have a `Requires` line in their RPM package specifying their related settings to prevent these issues at build time.

> network = "v1"

Reviewer 2: Would this also be possible (i.e. validating only a subsection of a top-level setting?)

```toml
# auditor.toml -- setting extension config file for `settings.auditor`
[extension.validates]
network.foo.bar = "v1"
```

Author: With this design it's not possible. The setting extension itself would have to determine whether or not relevant subtrees of other settings are changed.

Failure to populate a version listed as "supported" will result in a failed transaction.
See the next section about migrations for more details.

### Extension Versioning and Migrations
Bottlerocket's settings data model changes frequently: sometimes to support new features, or sometimes to correct mistaken model shapes.
In order to support changes to the settings model, settings must be capable of exposing versioning information about the settings extension, as well as providing faculties for migrating settings data to new versions.
Our versioning scheme will take heavy influence from Kubernetes' scheme for [versioning CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/).
Settings extension versioning information is surfaced through the extension config file:

```toml
# hostname.toml -- setting extension config file for settings.hostname
[extension]
supported-versions = [
     "v1",
     "v2"
]
default-version = "v2"
```

Any setting can be written or read at any of the `supported-versions` listed by the extension.

> Unversioned requests to read or write this setting are assumed to be the `default-version` and should only take place when a user makes a request using the CLI -- all automated process *should* refer to an explicit version.

Reviewer: Unversioned requests will also be used when handling user-data, right?

Author: Yes!

During a `set` transaction, the `apiserver` will consider the provided version to be the `canonical-version`, and will ask the settings extension to perform migrations from the `canonical` version to all supported versions using the `migrate` command.
These migrations always begin at the `canonical` version, and settings extensions are free to perform the migrations however they wish; however, Bottlerocket's Settings SDK will provide utilities to help settings developers support a series of linear setting upgrades and downgrades.

> Read on for more information about the Settings SDK

Reviewer: ```suggestion
Read on for more information about the Settings SDK.
```

All migrations are performed and stored at setting-write-time because Bottlerocket rollbacks may result in running an OS with a settings extension which does not support the version at which a given setting is stored.

> If Bottlerocket boots and finds that a setting is stored at a value which is not supported by it's installed setting extension, it will allow that version to persist unless a new value is `set` by the `apiserver`, at which point it is removed -- this is to attempt to prevent dataloss in the case that the image can eventually update to a new image which supports the stored setting version.

Reviewer: ```suggestion
If Bottlerocket boots and finds that a setting is stored at a value which is not supported by its installed setting extension, it will allow that version to persist unless a new value is `set` by the `apiserver`, at which point it is removed -- this is to attempt to prevent data loss in the case that the image can eventually update to a new image which supports the stored setting version.
```

Because settings are discovered at runtime on the system, and indeed may no longer be handled by models written in Rust, it will no longer be feasible to create an `apiclient` with the same strong typing as exists today.
It will instead be preferred that system services, including first-party services, all route their fetching of API settings through their own configuration files, which will be rendered by Bottlerocket whenever the related settings are changed.

Today, configuration templates are rendered by `schnauzer` using [handlebars](https://handlebarsjs.com/) templates.
`schnauzer` fetches the entirety of the settings tree, defines some useful helper functions, and then uses these as context for rendering the template.
With the introduction of settings extensions, this will change such that templates specify `(name, version)` tuples for all extensions needed to render the template.
These extension specifications will be given within the template as "front matter" -- a TOML block specified at the head of the document and delineated with `+++` as such:

```toml
+++
[required-extensions]
hostname = v1
kubernetes = v1

> +++

Reviewer: I still have some doubts about the "front matter" approach, but I'm willing to defer judgment until I see the implementation 😀 

I'm still somewhat partial to the filesystem-oriented approach, e.g.:
```
usr/
└── share
    └── templates
        └── etc
            └── kubelet
                ├── kubelet.config
                └── kubelet.config.required.d
                    ├── kubernetes
                    └── network
```

Since this is a bit more consistent with some of the other ways dependencies are modeled, at the cost of giving up tight binding to the template.

Reviewer 2: My take was something like this, where the metadata file is found by naming convention:

```
usr/
└── share
    └── templates
        └── etc
            └── kubelet
                ├── kubelet.config
                └── kubelet.config.info.toml
```

Author: I'll open another PR discussing particulars about this design.

# Below here the template is written as usual
```

This will allow Bottlerocket to find the owning extensions for the data needed to render the template.

> It also tightly binds the template to a given implementation of the requisite settings, making it much more difficult to accidentally break "downstream" templates with "usptream" changes.

Reviewer: ```suggestion
It also tightly binds the template to a given implementation of the requisite settings, making it much more difficult to accidentally break "downstream" templates with "upstream" changes.
```

> `schnauzer` can also continue to be invoked as a setting generator without the need for a templated configuration file.

Reviewer 2: Is this considered a temporary arrangement?

Author: I think this is actually just wrong, settings generators in concept go away. What may be relevant is that you could invoke schnauzer to arbitrarily render a view of Bottlerocket settings without needing to invoke the API.

```bash
$ schnauzer --required-extension 'hostname=v1' '{{ network.hostname }}'
```

#### Extending Handlebars Helpers
Handlebars has a few utilities for applying simple processes to data in order to format it appropriately for your document; however, you often need to implement your own "helpers" for more complex text formatting.
This is currently done in Bottlerocket by adding helper implementations to the `schnauzer` tool directly.
Bottlerocket settings extensions will expose versioned handlebar helpers:

```bash
# hostname.toml setting extension config file
[templating.v1] # template helpers are versioned
helpers = [
    "localhost_aliases"
]

$ ./hostname proto1 template-helper --version v1 --helper localhost_aliases 'arg1' 'arg2'
#  rendered output returned as a string on stdout
```

Within the handlerbars template, the helper name will be prepended with `$EXTENSION_NAME.` to avoid collisions with other settings extensions.
As an example, to use the `localhost_aliases` helper above from a template, your template may look something like this:

```

> {{ hostname.localhost_aliases(arg1, arg2) }}

Reviewer 3: Perhaps I missed it, but how are the helpers registered with handlebars?

Author: The algorithm is something like:

* The template rendering system will use the frontmatter section of the template to determine which settings extensions it needs to use to render the document.
* The template system uses the API to determine what handlebar helpers are provided by the settings extensions required.
* For each helper discovered, the template renderer will register a helper where the implementation results in an API call to execute the settings extension's method for calling that helper.

```

### Settings Generation

> Bottlerocket settings currently have the ability to compute a default value via something called a "settings generator".

Reviewer 3: I wouldn't say that computing a default value is settings generator's main purpose.  Their main purpose is to calculate settings at runtime for things that can't be given a static default.

Settings generation often depends on the presence of other settings.
It's also possible for settings to have interdependencies on generation order, for example:

* `foo.bar` depends on `baz.bot`
* `boz.blop` depends on `foo.bar`

In order to resolve this, settings generators are allowed to require that they be provided whatever data is currently applied to dependent settings.
The generators are "re-entrant" in that they return a status dictating whether or not they are finished generating, or if they need more data to proceed.
Bottlerocket will iteratively invoke generators until all generation has completed, or in cases where it seems likely that a deadlock has occurred.

```toml
# setting.toml requires the network and kubernetes settings before it can emit its own
[generation.required]
network = v1
kubernetes = v2
```

## Settings Extension Development and the Settings SDK
Settings extensions can be written in any programming language, so long as a symlink to a binary speaking the "extension protocol" lands in the right spot on the filesystem.

> That said, many settings extensions need to tackle the same challenges: How do we represent and implement settings versions? Migrations? Validation? Template utilities? Rather than forcing all extensions to implement this in isolation, we will create a shared library, called the "Settings SDK", which will be built in Bottlerocket's monorepo and published to crates.io.

Reviewer: Let's not name this or say what repo it will be built from in this doc; it's not clear that we'd put this crate in the core repo vs. having another dedicated project for it.

I would like to avoid having a wide variety of different artifacts published from a single repo, since that complicates CI/CD workflows and the releases page.

    ])
    .run_extension()
    .context("Settings extension encountered an error.")
}
```

## Asked and Anticipated Questions
### Why not use some existing schema or modeling language to define the shape of settings?
Existing modeling languages are powerful, but even in our existing Bottlerocket settings we have validations which we have implemented as functions in the Rust code (or otherwise wished we had a simple way to do so.) Rather than strip power from settings extensions by dictating that they use one of these modeling languages, we leave it entirely up to the binary to decide how best to validate inputs.
In many cases, this may mean that the binaries carry with them their own implementation of a popular schema language, like JSONschema.

### How will the `apiserver` handle concurrent requests in the face of transactional writes?
The `apiserver` will use a [filesystem-based](https://man7.org/linux/man-pages/man2/flock.2.html) reader-writer lock to ensure that the datastore is only accessed by concurrent readers or a single writer. Writes which are sent to the `apiserver` while the file is lock are placed in a queue.

## Appendices

> ### A. Assisted EC2 Userdata Generation and Validation

Reviewer: Would it be better to tackle this in a follow-up design? I agree that it's a potential benefit, but it seems orthogonal to the main concepts here.

Author: Totally fair. It's included here to capture some verbal discussion of this document that occurred and I wanted to capture it as a possible feature of the new system, but it can be delegated to another doc.

Bottlerocket user issues are often-enough caused by minor misuse of specifying Bottlerocket settings via ec2 Userdata.
The feedback loop here is relatively slow: you must launch a new Bottlerocket instance and determine whether or not it booted into the appropriate state in order to determine whether or not a given userdata document is correct.
Indeed, even Bottlerocket core developers wish for tooling to better generate and validate EC2 userdata without the need to interact with the EC2 APIs.

This feature is possible even in a world where settings are defined by extensions which are discovered dynamically.

> During a variant build, we can modify sodapop to extract the settings extension binaries and construct a Docker container which can use the information exposed by those binaries in order to validate userdata, or render a userdata template.

Reviewer: ```suggestion
During a variant build, we can modify twoliter to extract the settings extension binaries and construct a Docker container which can use the information exposed by those binaries in order to validate userdata, or render a userdata template.
```

The name of an extension is dictated by the name of the symlink placed into the extensions directory.
This name *must* correspond to the name of the configuration file for the extension.

Furthermore, every top-level setting in Bottlerocket's API must be owned by a singular extension.
This means that whenever the system must interact with a setting (be it via user action on the API, or rendering templates to be used as configuration files), Bottlerocket can identify the setting extension to invoke in order to satisfy that interaction.
Ownership is also dictated by the extension name defined by the symlink.

As an example, the following suggests the filesystem layout for a settings extension for the `settings.network` settings:

```
usr/
└── lib
    └── bottlerocket
        └── settings
            ├── config.d
            │   └── network.toml
            └── extensions.d
                └── network -> /usr/bin/network-extension
```

### Setting Storage
Today, Bottlerocket's `apiserver` utilizes models to perform strict validation of settings inputs, which are then serialized to JSON and stored in Bottlerocket's datastore.

> Writes to Bottlerocket settings is going to change in many ways under our new paradigm.

Reviewer 2: nit

```suggestion
Writing to Bottlerocket settings is going to change in many ways under our new paradigm.
```

or

Writes to Bottlerocket settings are going to change in many ways under our new paradigm.

#### Transactional Writes and Cross-Validation
Bottlerocket today is missing a mechanism for cross-validation of settings.
As an example, suppose we want to represent a range of integers in settings by allowing a `setting.min-value` and `setting.max-value` to be set -- it should be possible for settings extensions to set both simultaneously, while also verifying that the state from setting both is valid (`setting.min_value <= setting.max_value`).

Let us cast this requirement in a different light: imagine an OS administrator wishes to restrict what values can be written to an existing setting -- for example, they may wish to only allow settings values to `network.hosts` so long as they include as a subset the entries mandated by their organization.
To allow this, settings extensions will be capable of register themselves as a *validator* for any other setting on the host, including settings that they do not own.
When handling write requests for multiple settings in a single transaction, the `apiserver` will first gather the resulting writes that would occur under each settings extension involved in the writes.

> Afterwards, the `apiserver` will provide the provisional settings state to each settings extension which has been registered as a validator of any of the target settings.

Reviewer 2: ```suggestion
Then the `apiserver` will provide the provisional settings state to each settings extension which has been registered as a validator of any of the target settings.
```

Any one settings extension has the power to halt the transaction by returning a non-zero exit-code on validation.

Validations must specify the setting which triggers the validation, as well as a list of templates which will be rendered and provided to the validator.

```toml
# auditor.toml -- setting extension config file for `settings.auditor`
[[extension.validations]]

> validates = "network.hostname"

Reviewer 2: ```suggestion
validates = [ "network.hostname" ]
```

with-templates = [

>     "/templates/auditor-extension/validator-template.toml"

Reviewer 2: I'm not sure I understand how a template is involved in the validation.

See the next section about migrations for more details.

### Extension Versioning and Migrations
Bottlerocket's settings data model changes frequently: sometimes to support new features, or sometimes to correct mistaken model shapes.
In order to support changes to the settings model, settings must be capable of exposing versioning information about the settings extension, as well as providing faculties for migrating settings data to new versions.
Our versioning scheme will take heavy influence from Kubernetes' scheme for [versioning CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/).
Settings extension versioning information is surfaced through the extension config file:

```toml
# hostname.toml -- setting extension config file for settings.hostname
[extension]
name = "hostname",
supported-versions = [
     "v1",
     "v2"
]
default-version = "v2"
```

Any setting can be written or read at any of the `supported-versions` listed by the extension.
Unversioned requests to read or write this setting are assumed to be the `default-version`.

> During a `set` transaction, the `apiserver` will consider the provided version to be the `canonical-version`, and will ask the settings extension to perform migrations from the `canonical` version to all supported versions using the `migrate` command.

Reviewer 2: Should `migrate` be rolled into `set`? i.e. should `set` also do the "migrations" since there is no case when `set` would *not* be followed by `migrate`?

First-party tools have no such constraints, and so there exists tools that take either approach.

Because settings are discovered at runtime on the system, and indeed may no longer be handled by models written in Rust, it will no longer be feasible to create an `apiclient` with the same strong typing as exists today.
It will instead be preferred that system services, including first-party services, all route their fetching of API settings through their own configuration files, which will be rendered by Bottlerocket whenever the related settings are changed.

Today, configuration templates are rendered by `schnauzer` using [handlebars](https://handlebarsjs.com/) templates.
`schnauzer` fetches the entirety of the settings tree, defines some useful helper functions, and then uses these as context for rendering the template.
With the introduction of settings extensions, this will change such that templates specify `(name, version)` tuples for all extensions needed to render the template.
These extension specifications will be given within the template as "front matter" -- a TOML block specified at the head of the document and delineated with `---` as such:

```toml

> ---

Reviewer 4: I would suggest delimiting front mater with something other than triple dash.

Triple plus (`+++`) is my preference as [Hugo, Forestry, and Zola use it. Some other software uses `---toml`](https://github.com/denoland/deno_std/issues/3094#issuecomment-1400018219) but that is not symmetrical, the ending being triple dash.. yuck. 

Triple dash is likely to confuse people as that alone is widely used for yaml.

Reviewer 2: I would prefer not to use frontmatter since I'm concerned it will confuse syntax highlighters, linters, etc. Case in point, see the red highlights in this GitHub render. It rather see a separate file with a naming convention, theconfig.meta.toml or theconfig.frontmatter.toml... something like that.

[required-extensions]
hostname = v1
kubernetes = v1
---
# Below here the template is written as usual
```

This will allow Bottlerocket to find the owning extensions for the data needed to render the template.
It also tightly binds the template to a given implementation of the requisite settings, making it much more difficult to accidentally break "downstream" templates with "usptream" changes.

`schnauzer` can also continue to be invoked as a setting generator without the need for a templated configuration file.
In those circumstances, you are still required to list your extension dependencies and their version, but these can be done as command line arguments:

```bash
$ schnauzer --required-extension 'hostname=v1' '{{ network.hostname }}'
```

> As an optimization, templates which render to `systemd` units will invoke a restart command to systemd

Reviewer 2: Some might not be running yet during the boot sequence.

#### Extending Handlebars Helpers
Handlebars has a few utilities for applying simple processes to data in order to format it appropriately for your document; however, you often need to implement your own "helpers" for more complex text formatting.
This is currently done in Bottlerocket by adding helper implementations to the `schnauzer` tool directly.
Bottlerocket settings extensions will expose versioned handlebar helpers:

```bash
# hostname.toml setting extension config file

> [templating.v1] # template helpers are versioned
> helpers = [
>     "localhost_aliases"
> ]

Reviewer 2: Something seems off here. It looks like there would be multiple versions of the schema by which you can specify which helpers exist.

Maybe this? I guess I'm not super clear on the versioning of helpers.

```suggestion
[templating.helpers] # template helpers are versioned
localhost_aliases = "v1"
```

Author: This is by design.

Each existing version of a setting would have its own helpers that it provides. During template rendering, we request a settings at a given version, we need to ensure that the template helpers at that version continue to exist and provide the same API.

In practice, this likely means that you'd specify more or less the same helpers at all versions, but this gives us a valve to move to new versions.

$ ./hostname proto1 template-helper --version v1 --helper localhost_aliases 'arg1' 'arg2'
#  rendered output returned as a string on stdout
```

Within the handlerbars template, the helper name will be prepended with `$EXTENSION_NAME.` to avoid collisions with other settings extensions.
As an example, to use the `localhost_aliases` helper above from a template, your template may look something like this:

```
{{ hostname.localhost_aliases(arg1, arg2) }}
```

### Settings Generation
Bottlerocket settings currently have the ability to compute a default value via something called a "settings generator".
During boot, settings extensions are given a chance to populate settings with default values that are then persisted, effectively determining the setting value for the lifetime of that instance.

> Settings generation often depends on the presence of other settings; therefore the invocation protocol will allow settings generation to accept rendered templates as input:

Reviewer 2: rendered templates instead of other settings? wouldn't be more convenient for me to say I need `aws.region` than to find a template which contains that value?

Author: This design made `aws.region` more opaque from the apiserver's perspective, so it wouldn't be possible for the server to resolve that. It's one of the reasons why we'd need to move system services to use template files. Basically anything that would request a setting would do so via a template.
